### PR TITLE
[OpenCL] Enable OpenCL for GPU tests

### PR DESCRIPTION
**Pull request by valmat07**
Enable OpenCL support for GPU tests.

### CODE REVIEW

**Comment by valmat07** [comment]
This PR enables OpenCL test in CI. For that:

- Added USE_OPENCL flag in the build gpu script 
- Added adreno target in the CI target
- Moved relay opencl texture tests in different location to support run them independently. This is necessary due to not all relay tests supports OpenCL CI configuration
- Now opencl texture tests working in fp32.

File: CMakeLists.txt [hunk 1]
```diff
@@ -798,4 +798,4 @@ if(${SUMMARIZE})
   print_summary()
 endif()
 
-dump_options_to_file("${TVM_ALL_OPTIONS}")
+dump_options_to_file("${TVM_ALL_OPTIONS}")
```
Comment (echuraev) [inline CMakeLists.txt:801]: No meaningful changes. Please revert it.

File: tests/cpp-runtime/opencl/run_gtests.cc [hunk 2]
```diff
@@ -40,7 +40,7 @@ TVM_REGISTER_GLOBAL("opencl.run_gtests").set_body([](TVMArgs args, TVMRetValue*
   argv.push_back(const_cast<char*>("opencl_run_gtests"));
 
   // add parsed arguments
-  for (int i = 0; i < parsed_args.size(); ++i) {
+  for (uint i = 0; i < parsed_args.size(); ++i) {
```
Comment (echuraev) [inline tests/cpp-runtime/opencl/run_gtests.cc:43]: `size_t` should be better than `uint`.

File: tests/python/contrib/test_opencl/test_run_gtests.py [hunk 3]
```diff
@@ -28,6 +29,7 @@
 # for example to run all "foo" tests twice and observe gtest output run
 # pytest -sv <this file> --gtests_args="--gtest_filter=*foo* --gtest_repeat=2"
 @tvm.testing.requires_opencl
+@pytest.mark.xfail(raises=AttributeError)
```
Comment (echuraev) [inline tests/python/contrib/test_opencl/test_run_gtests.py:32]: Why do we need to mark this test as xfail?

File: tests/python/relay/opencl_texture/test_conv2d_nchw_texture.py [hunk 4]
```diff
@@ -437,7 +422,7 @@ def test_conv2d_vgg16_winograd_4d():
...
     with open(stat_file, "w") as f:
         f.write(
-            '{"input": ["opencl -keys=adreno,opencl,gpu -device=adreno -max_num_threads=256", "conv2d_nchw_winograd_acc32.image2d", [["TENSOR", [1, 512, 28, 28], "float16"], ["TENSOR", [512, 512, 3, 3], "float16"], [1, 1], [1, 1, 1, 1], [1, 1], "float16"], {}], "config": {"index": 1591, "code_hash": null, "entity": [["auto_unroll_max_step", "ot", 4], ["tile_y", "sp", [-1, 1, 32]], ["tile_x", "sp", [-1, 4, 2]], ["tile_rc", "sp", [-1, 8]]]}, "result": [[0.0037244], 0, 7.06374192237854, 1653898629.7427933], "version": 0.2, "tvm_version": "0.8.dev0"}\n'
+            '{"input": ["opencl -keys=adreno,opencl,gpu -device=adreno -max_num_threads=256", "conv2d_nchw_winograd_acc32.image2d", [["TENSOR", [1, 512, 28, 28], "float32"], ["TENSOR", [512, 512, 3, 3], "float32"], [1, 1], [1, 1, 1, 1], [1, 1], "float32"], {}], "config": {"index": 1591, "code_hash": null, "entity": [["auto_unroll_max_step", "ot", 4], ["tile_y", "sp", [-1, 1, 32]], ["tile_x", "sp", [-1, 4, 2]], ["tile_rc", "sp", [-1, 8]]]}, "result": [[0.0037244], 0, 7.06374192237854, 1653898629.7427933], "version": 0.2, "tvm_version": "0.8.dev0"}\n'
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/test_conv2d_nchw_texture.py:425]: You can use here value of variable `dtype` and the same below.

File: tests/python/relay/opencl_texture/test_conv2d_nchw_texture.py [hunk 5]
```diff
@@ -496,7 +479,9 @@ def test_conv2d_winograd_conv():
 
 
 @tvm.testing.requires_opencl
-def test_residual_block():
+@tvm.testing.parametrize_targets("opencl -device=adreno")
+@pytest.mark.xfail
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/test_conv2d_nchw_texture.py:483]: Why this test marked as `xfail`? 
cc: @elvin-n

File: tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py [hunk 6]
```diff
@@ -313,10 +303,9 @@ def test_conv2d_inceptionv3_35_35_strides():
 
 
 @tvm.testing.requires_opencl
-def test_conv2d_resnet50_v2_nhwc_3c():
-    target = "opencl --device=adreno"
-    dtype = "float16"
-
+@tvm.testing.parametrize_targets("opencl -device=adreno")
+@pytest.mark.xfail(raises=tvm._ffi.base.TVMError)
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py:307]: Why some of the tests are marked as xfail? I thought that all of them should be passed.

File: tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py [hunk 7]
```diff
@@ -598,7 +576,7 @@ def test_conv2d_vgg16_winograd_4d():
...
     with open(stat_file, "w") as f:
         f.write(
-            '{"input": ["opencl -keys=adreno,opencl,gpu -device=adreno -max_num_threads=256", "conv2d_nhwc_winograd_acc32.image2d", [["TENSOR", [1, 28, 28, 512], "float16"], ["TENSOR", [3, 3, 512, 512], "float16"], [1, 1], [1, 1, 1, 1], [1, 1], "float16"], {}], "config": {"index": 1591, "code_hash": null, "entity": [["auto_unroll_max_step", "ot", 4], ["tile_y", "sp", [-1, 1, 32]], ["tile_x", "sp", [-1, 4, 2]], ["tile_rc", "sp", [-1, 8]]]}, "result": [[0.0037244], 0, 7.06374192237854, 1653898629.7427933], "version": 0.2, "tvm_version": "0.8.dev0"}\n'
+            '{"input": ["opencl -keys=adreno,opencl,gpu -device=adreno -max_num_threads=256", "conv2d_nhwc_winograd_acc32.image2d", [["TENSOR", [1, 28, 28, 512], "float32"], ["TENSOR", [3, 3, 512, 512], "float32"], [1, 1], [1, 1, 1, 1], [1, 1], "float32"], {}], "config": {"index": 1591, "code_hash": null, "entity": [["auto_unroll_max_step", "ot", 4], ["tile_y", "sp", [-1, 1, 32]], ["tile_x", "sp", [-1, 4, 2]], ["tile_rc", "sp", [-1, 8]]]}, "result": [[0.0037244], 0, 7.06374192237854, 1653898629.7427933], "version": 0.2, "tvm_version": "0.8.dev0"}\n'
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py:579]: Use `dtype`

File: tests/python/relay/opencl_texture/utils/adreno_utils.py [hunk 8]
```diff
@@ -71,7 +71,6 @@ def build_run_compare(
         tvm_mod_nchwc = gpu_preprocess(tvm_mod)
     else:
         tvm_mod_nchwc = tvm_mod
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/utils/adreno_utils.py:73]: Please, revert this change

**Review by echuraev — CHANGES_REQUESTED** [review]
Thank you for your PR! Please, rebase it to the latest mainline. And one question. Will we run our cpp opencl tests in the CI?

File: tests/python/contrib/test_opencl/test_run_gtests.py (hunk 3 above)
Comment (valmat07) [reply tests/python/contrib/test_opencl/test_run_gtests.py:32]: I couldn't enable opencl cpp tests due to problems with CI environment. For enabling openCL Gtests we need to get source of googletests but there are no sources in the docker image. I also tried to get source from git (you can check my some commits for this) but when we reinstall gtest lib, other CPP tests become broken. So I decided to mark this test as xfail.

File: tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py (hunk 6 above)
Comment (valmat07) [reply tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py:307]: Some test in `test_conv2d_nhwc_texture.py` marked as xfail because in CI we have nvidia opencl lib, and this tests fails in  the CI. Locally, on my nvidia device, i also have this problem, but when I tried to run tests on my adreno device, everything works fine.

**Comment by valmat07** [comment]
@echuraev I found better solutions to exclude tests, now they are excluded in CI scripts. Please check

File: tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py [hunk 9]
```diff
@@ -398,11 +384,9 @@ def test_conv2d_inceptionv3_nhwc_3c():
...
 
 @tvm.testing.requires_opencl
-def test_conv2d_1x1_16c16spatial():
-    target = "opencl --device=adreno"
-    dtype = "float16"
-
-    input_shape = (1, 256, 256, 16)
+@tvm.testing.parametrize_targets("opencl -device=adreno")
+def test_conv2d_1x1_16c16spatial(target, dtype):
+    input_shape = (1, 128, 128, 16)
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py:389]: Why did you change input shape?

File: tests/python/relay/opencl_texture/utils/adreno_utils.py [hunk 10]
```diff
@@ -71,7 +71,7 @@ def build_run_compare(
...
-
+
```
Comment (echuraev) [inline tests/python/relay/opencl_texture/utils/adreno_utils.py:74]: Remove the spaces

File: CMakeLists.txt (hunk 1 above)
Comment (echuraev) [inline CMakeLists.txt:801]: Revert this change.

**Review by echuraev — COMMENTED** [review]
Several minor comments

File: tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py (hunk 9 above)
Comment (valmat07) [reply tests/python/relay/opencl_texture/test_conv2d_nhwc_texture.py:389]: Because there was an error that there is not enough memory for opencl

**Comment by valmat07** [comment]
@tvm-bot rerun

**Comment by valmat07** [comment]
@tvm-bot rerun

**Review by echuraev — APPROVED** [review]
LGTM. Thanks!

File: tests/scripts/task_python_integration.sh [hunk 11]
```diff
@@ -61,12 +61,16 @@ run_pytest cython ${TVM_INTEGRATION_TESTSUITE_NAME}-dso_plugin_module-1 apps/dso
...
-run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib tests/python/contrib --ignore=tests/python/contrib/test_ethosu --ignore=tests/python/contrib/test_cmsisnn
+run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib tests/python/contrib --ignore=tests/python/contrib/test_ethosu --ignore=tests/python/contrib/test_cmsisnn --ignore=tests/python/contrib/test_opencl
 
 # forked is needed because the global registry gets contaminated
 TVM_TEST_TARGETS="${TVM_RELAY_TEST_TARGETS:-llvm;cuda}" \
     run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-relay tests/python/relay
 
+# OpenCL texture test. Deselected specific tests that fails  in CI
+TVM_TEST_TARGETS="${TVM_RELAY_OPENCL_TEXTURE_TARGETS:-opencl}" \
+    run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-opencl-texture tests/python/relay/opencl_texture --deselect=tests/python/relay/opencl_texture/test_conv2d_nchw_texture.py::test_residual_block
```
Comment (driazati) [inline tests/scripts/task_python_integration.sh:72]: Instead of deselecting here can you:
1. File a GitHub issue with the test name and some context on why it fails with OpenCL
2. Move the skipping to the test itself via [`pytest.mark.skipIf`](https://docs.pytest.org/en/7.1.x/how-to/skipping.html) and be sure to add a `reason="See <link to apache/tvm GitHub issue>"`? 

That might make it so you don't need to change this script at all

Comment (valmat07) [reply tests/scripts/task_python_integration.sh:72]: But the main thing here is that this test works well locally on adreno devices, so you don’t want to skip it. But in CI environment due to we have nvidia device so different libOpenCL, this test doesn't work properly.

Comment (valmat07) [reply tests/scripts/task_python_integration.sh:72]: Generally we want to exclude some tests only for nvidia devices, but I couldn't find the proper way to exclude tests for nvidia. So maybe there is some way to exclude tests by making sure it doesn't run on nvidia?

Comment (driazati) [reply tests/scripts/task_python_integration.sh:72]: You can do whatever logic is necessary in the condition for the `skipIf` so that it's only skipped in the specific cases you want (e.g. here maybe just `skipIf(tvm.testing.utils.IS_IN_CI, reason="...")`). Also in `tvm/testing/utils.py` take a look at `requires_cuda` or `uses_gpu` as they may already do what you need

**Comment by valmat07** [comment]
@tvm-bot rerun

File: tests/scripts/task_python_integration.sh (hunk 11 above)
Comment (valmat07) [reply tests/scripts/task_python_integration.sh]: Added skipif instead of deselect. Thanks for the suggestion!

File: tests/scripts/task_python_integration.sh [hunk 12]
```diff
@@ -61,12 +61,14 @@ run_pytest cython ${TVM_INTEGRATION_TESTSUITE_NAME}-dso_plugin_module-1 apps/dso
...
-run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib tests/python/contrib --ignore=tests/python/contrib/test_ethosu --ignore=tests/python/contrib/test_cmsisnn
-
+run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-contrib tests/python/contrib --ignore=tests/python/contrib/test_ethosu --ignore=tests/python/contrib/test_cmsisnn 
 # forked is needed because the global registry gets contaminated
 TVM_TEST_TARGETS="${TVM_RELAY_TEST_TARGETS:-llvm;cuda}" \
     run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-relay tests/python/relay
 
+# OpenCL texture test. Deselected specific tests that fails  in CI
+TVM_TEST_TARGETS="${TVM_RELAY_OPENCL_TEXTURE_TARGETS:-opencl}" \
+    run_pytest ctypes ${TVM_INTEGRATION_TESTSUITE_NAME}-opencl-texture tests/python/relay/opencl_texture
```
Comment (driazati) [inline tests/scripts/task_python_integration.sh:71]: one more thing, won't this get run as part of `test/python/relay` in the `run_pytest` above this one?

File: tests/python/contrib/test_opencl/test_run_gtests.py [hunk 13]
```diff
@@ -28,6 +28,7 @@
 # for example to run all "foo" tests twice and observe gtest output run
 # pytest -sv <this file> --gtests_args="--gtest_filter=*foo* --gtest_repeat=2"
 @tvm.testing.requires_opencl
+@pytest.mark.skipif(True, reason="failed due to nvidia libOpencl in the CI")
```
Comment (driazati) [inline tests/python/contrib/test_opencl/test_run_gtests.py:31]: did you want to condition this? with `True` it will always be skipped, even locally

File: tests/scripts/task_python_integration.sh (hunk 12 above)
Comment (valmat07) [reply tests/scripts/task_python_integration.sh:71]: a large number of tests in `tests/python/realy/` for OpenCL do not run, for various reasons, so I took out the texture tests separately so as not to mark relay tests for some special occasions.

**Comment by TejashShah** [comment]
cc @masahi @junrushao, it needs a merge trigger!
